pkg/controllers/cstorvolumeconfig/controller_test.go

```diff
@@ -17,16 +17,341 @@ limitations under the License.
+}
+
+func (f *fixture) run(cvcName string) {
+	testConfig := testConfig{
+		loopCount: 1,
+		loopDelay: time.Second * 0,
+	}
+	f.run_(cvcName, true, false, testConfig)
+}
+
+func (f *fixture) run_(cvcName string, startInformers bool, expectError bool, testConfig testConfig) {
+	c, informers, recorder, err := f.newCVCController()
+	if err != nil {
+		f.t.Fatalf("error creating cspc controller: %v", err)
+	}
+	if startInformers {
+		stopCh := make(chan struct{})
+		defer close(stopCh)
+		informers.Start(stopCh)
+	}
+
+	defer func(recorder *record.FakeRecorder) {
+		close(recorder.Events)
+	}(recorder)
+
+	go printEvent(recorder)
+
+	for i := 0; i < testConfig.loopCount; i++ {
+		err = c.syncHandler(cvcName)
+		if !expectError && err != nil {
+			f.t.Errorf("error syncing cvc: %v", err)
+		} else if expectError && err == nil {
+			f.t.Error("expected error syncing cvc, got nil")
+		}
+		if testConfig.loopCount > 1 {
+			time.Sleep(testConfig.loopDelay)
+		}
+	}
+}
+
+// printEvent prints the events reported by controller
+func printEvent(recorder *record.FakeRecorder) {
+	rocket := html.UnescapeString("&#128640;")
+	warning := html.UnescapeString("&#10071;")
+	for {
+		msg, ok := <-recorder.Events
+		// Channel is closed
+		if !ok {
+			break
+		}
+		if strings.Contains(msg, "Normal") {
+			// Below line prints 🚀 to identify event
+			fmt.Println("Event:  ", rocket, msg)
+		} else {
+			// Below line prints ❗ to identify event
+			fmt.Println("Event:  ", warning, msg)
+		}
+	}
+}
+
+// Returns 0 for resyncPeriod in case resyncing is not needed.
+func NoResyncPeriodFunc() time.Duration {
+	return 0
+}
+
+// newCVCController returns a fake cvc controller
+func (f *fixture) newCVCController() (*CVCController, openebsinformers.SharedInformerFactory, *record.FakeRecorder, error) {
+	cvcInformerFactory := openebsinformers.NewSharedInformerFactory(f.openebsClient, NoResyncPeriodFunc())
+
+	// Build a fake controller
+	c := NewCVCControllerBuilder().
+		withOpenEBSClient(f.openebsClient).
+		withKubeClient(f.k8sClient).
+		withCVCLister(cvcInformerFactory).
+		withCVLister(cvcInformerFactory).
+		withCVRLister(cvcInformerFactory).
+		withEventHandler(cvcInformerFactory).
+		withWorkqueueRateLimiting().
+		CVCController
+	recorder := record.NewFakeRecorder(1024)
+	c.recorder = recorder
+	c.cvcSynced = alwaysReady
+	c.cvrSynced = alwaysReady
+	c.enqueueCVCConfig = c.enqueueCVC
+
+	for _, cspi := range f.cspiLister {
+		cvcInformerFactory.Cstor().V1().CStorPoolInstances().Informer().GetIndexer().Add(cspi)
+	}
+
+	for _, cv := range f.cvLister {
+		cvcInformerFactory.Cstor().V1().CStorVolumes().Informer().GetIndexer().Add(cv)
+	}
+	for _, cvc := range f.cvcLister {
+		cvcInformerFactory.Cstor().V1().CStorVolumeConfigs().Informer().GetIndexer().Add(cvc)
+	}
+
+	for _, cvr := range f.cvrLister {
+		cvcInformerFactory.Cstor().V1().CStorVolumeReplicas().Informer().GetIndexer().Add(cvr)
+	}
+
+	return c, cvcInformerFactory, recorder, nil
```

Comment:
nit: Instead of returning `recorder` we can typecast recorder interface in controller struct `c.recorder.(*record.FakeRecorder)`

Reply:
yes, i will change that in upcoming PRs with some new tests